### PR TITLE
Always monitor the CNI config dir.

### PR DESCRIPTION
**Pull request by ijc**
Even if a valid configuration is found at start of day we still need to watch
for updates to that config.

This is best achieved (to avoid racing with the watcher) by moving the initial
sync into `monitorNetDir`.

Discovered via https://github.com/containerd/cri-containerd/issues/545.

Signed-off-by: Ian Campbell <ijc@docker.com>

### CODE REVIEW

**Comment by ijc** [comment]
With this a potential future cleanup might be to make `monitorNetDir` a normal synchronous function which returns an error rather than a goroutine entry, which would allow it to return the various errors which might occur during initialisation up the call chain instead of logging and swallowing. In this case `plugin.monitorNetDirChan` might also become redundant.

An issue related to the problem fixed here, which I haven't tackled, is what happens if the monitoring thread fails, at present things will continue to use the last seen configuration and ignore any future updates.

File: pkg/ocicni/ocicni.go [hunk 1]
```diff
@@ -133,10 +133,26 @@ func (plugin *cniNetworkPlugin) monitorNetDir() {
 	}()
 
 	if err = watcher.Add(plugin.pluginDir); err != nil {
-		logrus.Error(err)
+		logrus.Errorf("Failed to add watch on %q: %v", err)
```
Comment (Random-Liu) [inline pkg/ocicni/ocicni.go]: `%q`?

File: pkg/ocicni/ocicni.go [hunk 2]
```diff
@@ -167,12 +183,7 @@ func InitCNI(pluginDir string, cniDirs ...string) (CNIPlugin, error) {
 		return nil, err
 	}
 
-	if err := plugin.syncNetworkConfig(); err != nil {
-		// We do not have a valid default network, so start the
-		// monitoring thread.  Network setup/teardown requests
-		// will fail until we have a valid default network.
-		go plugin.monitorNetDir()
-	}
+	go plugin.monitorNetDir()
```
Comment (Random-Liu) [inline pkg/ocicni/ocicni.go]: I believe it is a bug initially. `monitorNetDir` doesn't return after `syncNetworkConfig`, so I believe we want it to keep syncing network config. But we don't always start the loop.

Actually in cni in kubernetes, we do keep syncing network config https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/network/cni/cni.go#L200.

Given so, I think we should accept this change.

**Comment by Random-Liu** [comment]
@mrunalp @runcom

**Comment by runcom** [comment]
Makes sense (please address Lantao s comments)

**Comment by ijc** [comment]
Fixed the format string.

**Comment by mrunalp** [comment]
@dcbw @rajatchopra PTAL

**Comment by Random-Liu** [comment]
@dcbw @rajatchopra Could you take a look?

**Comment by Random-Liu** [comment]
@rajatchopra Any concern about this?

**Comment by rajatchopra** [comment]
Looks good to me. The idea is valid. But maybe we can do better than relying on the plugin.Lock! One failing condition is that we may set a stale configuration. i.e when the initial sync picks up a config, but before it sets the plugin, an event changes the config that the monitor thread catches. The monitor callback sets the plugin, and then finally the initial sync code gets scheduled and it sets the plugin based on the config prior to the event.
The result would be a plugin based on stale config files.

Will review this again tomorrow. Maybe I understood this wrong on a mobile phone.

**Comment by ijc** [comment]
@rajatchopra I think you are right, this sort of thing is why I filed https://github.com/fsnotify/fsnotify/issues/236 ;-) I'll take a look today, either moving the goroutine launch or adding some explicit interlock ought to do the trick I think.

**Comment by ijc** [comment]
@rajatchopra I think starting the goroutine later as I've just done closes the whole you spotted.

Thanks, Ian.

**Pull request merged**